File: benchmark_python/postgres/worker.py
"""PostgreSQL insert worker and query worker (asyncpg)."""
from __future__ import annotations

import asyncio
import logging
import os
import time
from typing import Any

from ..base_worker import (
    BaseAsyncInsertWorker,
    _mrns_from_batch,
    _split_originals_duplicates,
)
from ..config import QUERY_SENTINEL
from . import backend

logger = logging.getLogger(__name__)

DEFAULT_HOST = "localhost"
DEFAULT_PORT = 5432
# When pgbouncer is enabled, connect to pgbouncer (not Postgres directly).
DEFAULT_PGBOUNCER_HOST = "pgbouncer"
DEFAULT_PGBOUNCER_PORT = 6432


class PostgresWorker(BaseAsyncInsertWorker):
    """PostgreSQL worker: context (setup/teardown/pools) and insert worker in one. Uses parameterized INSERT; when pgbouncer_enabled, query hint /* pgbouncer.database = db */ prepended to INSERT in one statement."""

    def __init__(self) -> None:
        self.insert_pool = None
        self.select_pool = None
        self.pgbouncer_enabled = False
        # BaseAsyncInsertWorker attributes (set in make_worker before run() is used)
        self.insertion_queue = None
        self.query_queue = None
        self.inserted_lock = None
        self.inserted_shared = None
        self.batch_size = 0
        self.queries_per_record = 1

    async def setup(
        self,
        num_workers: int,
        target_rps: int,
        init_schema: bool = True,
        pgbouncer_enabled: bool = False,
    ) -> PostgresWorker:
        """Create asyncpg pools, prewarm, optionally init schema. Returns self."""
        if self.insert_pool is not None:
            raise RuntimeError("PostgresWorker.setup() already called")
        self.pgbouncer_enabled = pgbouncer_enabled
        if pgbouncer_enabled:
            host = os.environ.get("POSTGRES_PGBOUNCER_HOST") or DEFAULT_PGBOUNCER_HOST
            port = int(os.environ.get("POSTGRES_PGBOUNCER_PORT") or str(DEFAULT_PGBOUNCER_PORT))
            database = "postgres1"
        else:
            host = os.environ.get("POSTGRES_HOST") or DEFAULT_HOST
            port = int(os.environ.get("POSTGRES_PORT") or str(DEFAULT_PORT))
            database = None
        logger.info(
            "Creating PostgreSQL pools at %s:%d (%d insert + %d select)%s ...",
            host, port, num_workers, num_workers,
            " (pgbouncer alternate postgres1/postgres2 by batch index)" if pgbouncer_enabled else "",
        )
        skip_session_set = pgbouncer_enabled
        pool_kw: dict[str, Any] = {"database": database} if database else {}
        self.insert_pool = await backend.create_pool(host, port, num_workers, **pool_kw)
        await backend.prewarm_pool(self.insert_pool, num_workers, skip_session_set=skip_session_set)
        self.select_pool = await backend.create_pool(host, port, num_workers, **pool_kw)
        await backend.prewarm_pool(self.select_pool, num_workers, skip_session_set=skip_session_set)
        if init_schema:
            conn = await self.insert_pool.acquire()
            try:
                await backend.init_schema(conn)
            finally:
                await self.insert_pool.release(conn)
        logger.info("Starting insertions (target %d rows/sec) ...", target_rps)
        return self

    async def teardown(self) -> None:
        if self.select_pool is not None:
            await self.select_pool.close()
            self.select_pool = None
        if self.insert_pool is not None:
            await self.insert_pool.close()
            self.insert_pool = None

    def make_worker(
        self,
        insertion_queue: asyncio.Queue[tuple[str, list] | None],
        query_queue: asyncio.Queue,
        inserted_lock: asyncio.Lock,
        inserted_shared: list[float],
        batch_size: int,
        queries_per_record: int = 1,
        pgbouncer_enabled: bool = False,
    ) -> PostgresWorker:
        """Set queue/state for run() and return self (this object is the insert worker)."""
        self.insertion_queue = insertion_queue
        self.query_queue = query_queue
        self.inserted_lock = inserted_lock
        self.inserted_shared = inserted_shared
        self.batch_size = batch_size
        self.queries_per_record = queries_per_record
        return self

    async def get_max_patient_counter(self) -> int:
        conn = await self.select_pool.acquire()
        try:
            return await backend.get_max_patient_counter(conn)
        finally:
            await self.select_pool.release(conn)

    async def get_connection(self) -> Any:
        return await self.insert_pool.acquire()

    async def release_connection(self, conn: Any) -> None:
        await self.insert_pool.release(conn)

    async def _flush(self, batch: list, query_hint: str = "") -> None:
        """When pgbouncer is enabled, use a separate connection per sub-batch (originals,
        then duplicates) so only one hint + INSERT runs per connection.
        query_hint is the prepared hint string set by the producer.
        """
        if not batch:
            return
        if not self.pgbouncer_enabled:
            await super()._flush(batch, query_hint)
            return
        originals, duplicates = _split_originals_duplicates(batch)
        async with self.inserted_lock:
            self.inserted_shared[5] += 1
        total_rows = 0
        total_latency_sec = 0.0
        n_statements = 0
        try:
            if originals:
                conn = await self.get_connection()
                try:
                    rows_db = [(r[0], r[1], r[2]) for r in originals]
                    t0 = time.perf_counter()
                    n, stmts, db_used = await self.insert_batch(conn, rows_db, query_hint)
                    total_rows += n
                    total_latency_sec += time.perf_counter() - t0
                    n_statements += stmts
                    if db_used:
                        async with self.inserted_lock:
                            if db_used == backend.PGBOUNCER_DB1:
                                self.inserted_shared[7] += n
                            else:
                                self.inserted_shared[8] += n
                finally:
                    await self.release_connection(conn)
            if duplicates:
                conn = await self.get_connection()
                try:
                    rows_db = [(r[0], r[1], r[2]) for r in duplicates]
                    t0 = time.perf_counter()
                    n, stmts, db_used = await self.insert_batch(conn, rows_db, query_hint)
                    total_rows += n
                    total_latency_sec += time.perf_counter() - t0
                    n_statements += stmts
                    if db_used:
                        async with self.inserted_lock:
                            if db_used == backend.PGBOUNCER_DB1:
                                self.inserted_shared[7] += n
                            else:
                                self.inserted_shared[8] += n
                finally:
                    await self.release_connection(conn)
            n_originals = len(originals)
            n_duplicates = len(duplicates)
            async with self.inserted_lock:
                self.inserted_shared[0] += total_rows
                self.inserted_shared[1] += n_originals
                self.inserted_shared[2] += n_duplicates
                self.inserted_shared[3] += total_latency_sec
                self.inserted_shared[4] += n_statements
            if self.queries_per_record > 0:
                insert_time = time.time()
                for mrn in _mrns_from_batch(batch):
                    await self.query_queue.put((mrn, insert_time))
        finally:
            async with self.inserted_lock:
                self.inserted_shared[5] -= 1
                if self.inserted_shared[5] < 0:
                    self.inserted_shared[5] = 0

    async def insert_batch(
        self, conn: Any, batch: list[tuple[str, str, str]], query_hint: str = ""
    ) -> tuple[int, int, str | None]:
        """Returns (rows_inserted, statement_count, db_used). query_hint is the prepared hint string."""
        if self.pgbouncer_enabled and query_hint:
            n = await backend.insert_batch_with_pgbouncer_hint(conn, batch, query_hint)
            db_used = backend.database_from_query_hint(query_hint)
            return n, 1, db_used
        n = await backend.insert_batch(conn, batch)
        return n, 1, None


async def run_query_worker_postgres(
    query_queue: asyncio.Queue,
    pool: Any,
    queries_lock: asyncio.Lock,
    queries_shared: list[float],
    queries_per_record: int,
    query_delay_sec: float,
    query_rate_limiter: Any,
    ignore_select_errors: bool,
) -> None:
    while True:
        item = await query_queue.get()
        if item is QUERY_SENTINEL:
            return
        mrn, insert_time = item
        if query_delay_sec > 0:
            deadline = insert_time + query_delay_sec
            sleep_sec = deadline - time.time()
            if sleep_sec > 0:
                await asyncio.sleep(sleep_sec)
        conn = await pool.acquire()
        try:
            total_latency_sec = 0.0
            failed = 0
            for _ in range(queries_per_record):
                if query_rate_limiter is not None:
                    await query_rate_limiter.acquire()
                t0 = time.perf_counter()
                rows = await backend.query_by_primary_key(conn, mrn)
                total_latency_sec += time.perf_counter() - t0
                if len(rows) != 1:
                    failed += 1
                    if not ignore_select_errors:
                        logger.error(
                            "Query by primary key returned %d rows for MEDICAL_RECORD_NUMBER=%s (expected 1)",
                            len(rows), mrn,
                        )
            async with queries_lock:
                queries_shared[0] += queries_per_record
                queries_shared[1] += total_latency_sec
                queries_shared[2] += failed
        finally:
            await pool.release(conn)

File: benchmark-go/postgres/worker.go
package postgres

import (
	"context"
	"log"
	"os"
	"strconv"
	"sync/atomic"
	"time"

	"github.com/db-benchmarking/benchmark-go"
	"github.com/jackc/pgx/v5/pgxpool"
)

const defaultHost = "localhost"
const defaultPort = 5432
// When PgbouncerEnabled, connect to pgbouncer (not Postgres directly).
const defaultPgbouncerHost = "pgbouncer"
const defaultPgbouncerPort = 6432

const pgbouncerDB1 = "postgres1"
const pgbouncerDB2 = "postgres2"

// Backend holds the pool and implements benchmarkgo.InsertBackend.
// When pgbouncerMode is true, InsertBatch runs SET pgbouncer.database prepended to INSERT, flip-flopping postgres1/postgres2.
type Backend struct {
	pool            *pgxpool.Pool
	pgbouncerMode   bool
	pgbouncerUseDB1 atomic.Bool
}

// GetConn acquires a connection from the pool.
func (b *Backend) GetConn() interface{} {
	conn, err := b.pool.Acquire(context.Background())
	if err != nil {
		log.Printf("postgres Acquire: %v", err)
		return nil
	}
	return conn
}

// ReleaseConn returns the connection to the pool.
func (b *Backend) ReleaseConn(c interface{}) {
	if conn, ok := c.(*pgxpool.Conn); ok {
		conn.Release()
	}
}

// InsertBatch inserts rows using the given connection (must be *pgxpool.Conn). Returns (rowsInserted, statementCount, error).
// When pgbouncerMode is true, runs one round-trip: "SET pgbouncer.database = '...'; INSERT ..." (PgBouncer strips SET, forwards INSERT).
func (b *Backend) InsertBatch(conn interface{}, rows []benchmarkgo.RowForDB) (int, int, error) {
	c, ok := conn.(*pgxpool.Conn)
	if !ok {
		return 0, 0, nil
	}
	ctx := context.Background()
	if b.pgbouncerMode && len(rows) > 0 {
		useDB1 := b.pgbouncerUseDB1.Load()
		b.pgbouncerUseDB1.Store(!useDB1)
		db := pgbouncerDB1
		if !useDB1 {
			db = pgbouncerDB2
		}
		sql, args, err := BuildPgbouncerSetInsertStatement(rows, db)
		if err != nil {
			return 0, 0, err
		}
		if _, err := c.Exec(ctx, sql, args...); err != nil {
			return 0, 0, err
		}
		return len(rows), 1, nil
	}
	n, err := InsertBatch(ctx, c, rows)
	if err != nil {
		return n, 0, err
	}
	return n, 1, nil
}

// Context handles setup/teardown and query workers for PostgreSQL.
type Context struct {
	insertPool        *pgxpool.Pool
	selectPool        *pgxpool.Pool
	PgbouncerEnabled  bool
}

// Setup creates insert pool and optionally a separate select pool. When PgbouncerEnabled, uses one pool (postgres1) and SET pgbouncer.database with INSERT back-to-back.
func (c *Context) Setup(numWorkers, targetRPS int, queriesPerRecord int) (benchmarkgo.InsertBackend, error) {
	if c.insertPool != nil {
		log.Fatal("postgres Setup already called")
	}
	var host string
	var port int
	if c.PgbouncerEnabled {
		host = os.Getenv("POSTGRES_PGBOUNCER_HOST")
		if host == "" {
			host = defaultPgbouncerHost
		}
		port = defaultPgbouncerPort
		if p := os.Getenv("POSTGRES_PGBOUNCER_PORT"); p != "" {
			if v, err := strconv.Atoi(p); err == nil {
				port = v
			}
		}
	} else {
		host = os.Getenv("POSTGRES_HOST")
		if host == "" {
			host = defaultHost
		}
		port = defaultPort
		if p := os.Getenv("POSTGRES_PORT"); p != "" {
			if v, err := strconv.Atoi(p); err == nil {
				port = v
			}
		}
	}
	ctx := context.Background()
	if c.PgbouncerEnabled {
		log.Printf("Creating PostgreSQL connection pool at %s:%d (pgbouncer: postgres1, SET pgbouncer.database + INSERT flip-flop postgres1/postgres2, %d insert)",
			host, port, numWorkers)
		insertPool, err := CreatePoolWithDB(ctx, host, port, numWorkers, pgbouncerDB1)
		if err != nil {
			return nil, err
		}
		c.insertPool = insertPool
		if err := PrewarmPool(ctx, insertPool, numWorkers); err != nil {
			insertPool.Close()
			return nil, err
		}
		c.selectPool, _ = CreatePoolWithDB(ctx, host, port, numWorkers, pgbouncerDB1)
		if c.selectPool != nil {
			_ = PrewarmPool(ctx, c.selectPool, numWorkers)
		}
		if err := InitSchema(ctx, insertPool); err != nil {
			insertPool.Close()
			if c.selectPool != nil {
				c.selectPool.Close()
			}
			return nil, err
		}
		log.Printf("Starting insertions (target %d rows/sec) ...", targetRPS)
		be := &Backend{pool: insertPool, pgbouncerMode: true}
		be.pgbouncerUseDB1.Store(true)
		return be, nil
	}
	log.Printf("Creating PostgreSQL connection pool(s) at %s:%d (%d insert connections)",
		host, port, numWorkers)
	if queriesPerRecord > 0 {
		log.Printf("  + %d select connections for query workers", numWorkers)
	}
	insertPool, err := CreatePool(ctx, host, port, numWorkers)
	if err != nil {
		return nil, err
	}
	c.insertPool = insertPool
	if err := PrewarmPool(ctx, insertPool, numWorkers); err != nil {
		insertPool.Close()
		return nil, err
	}
	if queriesPerRecord > 0 {
		selectPool, err := CreatePool(ctx, host, port, numWorkers)
		if err != nil {
			insertPool.Close()
			return nil, err
		}
		c.selectPool = selectPool
		if err := PrewarmPool(ctx, selectPool, numWorkers); err != nil {
			insertPool.Close()
			selectPool.Close()
			return nil, err
		}
	}
	if err := InitSchema(ctx, insertPool); err != nil {
		insertPool.Close()
		if c.selectPool != nil {
			c.selectPool.Close()
		}
		return nil, err
	}
	log.Printf("Starting insertions (target %d rows/sec) ...", targetRPS)
	return &Backend{pool: insertPool}, nil
}

// Teardown closes all pools.
func (c *Context) Teardown() {
	if c.selectPool != nil {
		c.selectPool.Close()
		c.selectPool = nil
	}
	if c.insertPool != nil {
		c.insertPool.Close()
		c.insertPool = nil
	}
}

// GetMaxPatientCounter returns the max patient ordinal in the DB. Uses insert pool when select pool is not initialized.
func (c *Context) GetMaxPatientCounter() (int, error) {
	pool := c.selectPool
	if pool == nil {
		pool = c.insertPool
	}
	conn, err := pool.Acquire(context.Background())
	if err != nil {
		return -1, err
	}
	defer conn.Release()
	return GetMaxPatientCounter(context.Background(), conn)
}

// RunQueryWorker consumes from queryQueue, runs queries_per_record lookups per MRN, reports via benchmarkgo.AddQuery.
// workerIndex is the 0-based index of this query worker.
func (c *Context) RunQueryWorker(
	workerIndex int,
	queryQueue <-chan *benchmarkgo.QueryJob,
	queriesPerRecord int,
	queryDelaySec float64,
	ignoreSelectErrors bool,
) {
	_ = workerIndex // reserved for logging/tracing
	for job := range queryQueue {
		if job == nil {
			return
		}
		if queryDelaySec > 0 {
			deadline := job.InsertTime.Add(time.Duration(queryDelaySec * float64(time.Second)))
			if time.Now().Before(deadline) {
				time.Sleep(time.Until(deadline))
			}
		}
		conn, err := c.selectPool.Acquire(context.Background())
		if err != nil {
			continue
		}
		t0 := time.Now()
		var failed int
		for i := 0; i < queriesPerRecord; i++ {
			n, _ := QueryByPrimaryKey(context.Background(), conn, job.MRN)
			if n != 1 {
				failed++
				if !ignoreSelectErrors {
					log.Printf("Query by primary key returned %d rows for MEDICAL_RECORD_NUMBER=%s (expected 1)", n, job.MRN)
				}
			}
		}
		latencyMicros := time.Since(t0).Microseconds()
		conn.Release()
		benchmarkgo.AddQuery(int64(queriesPerRecord), latencyMicros, int64(failed))
	}
}
